### ISSUE TITLE
migrate thumbor to not use exif - rotation / aspect ratio is wrong

### ISSUE THREAD

**Issue by scd** [theCrag/website#2303]
Exif errors in user supplied topo images means that thumbor does not honor some exif orientation specfications. There is no way for us to tell the failure case. To get around this we have to switch thumbor to never use exif.

This is a fair bit of work and requires data migration:
- [x] every image with an exif will have to have their rotate field updated
- [ ] deduce an initial rotation from exif when image is uploaded
- [ ] turn off thumbor exif honouring
- [ ] remove the aspectFlipped variable and work out flipping the aspect from the rotate variable
- [ ] update aspect rotator function to only consider rotation, not exif. I don't think we need this rotator function any more but rather just do it in the template based on the rotate variable (rotate view port but not thumbor dimenisions).
- [ ] most important of all: have a rotate left / right button on the photo page. Let the owner do this, and also let anyone with coordinator role at that node do it

**Comment by brendanheywood** [comment]
Possibly another example of this to test it's fixed the issue:

* [ ] https://www.thecrag.com/climbing/canada/sullys-hangout

**Comment by brendanheywood** [comment]
Also there are heaps of busted one's in arapiles, not sure why exactly, but more examples to test afterwards:

* [ ] https://www.thecrag.com/climbing/australia/arapiles/atridae/photos

**Comment by scd** [comment]
From my investigation the reason for the rotation problems is that thumbor will not honor exif rotation if it finds an unrelated error in the exif format, however our system thinks that it will. There is no way we can predict this. The only way around is to abondon the exif rotation via thumbor and just use our system to specify the rotation. This is actually a fair bit of work in particular it will be a lot of work to test, because we need two versions of our image database.

**Comment by brendanheywood** [comment]
The deeper problem here is not that thumbor is messing stuff up, it's that we have zero ability to workaround it. I'd be fine right now to ignore the exif bugs and just add an easy way for either the owner or an admin to fix the rotation.

From my understanding of the bug, we have our system which reads the exif using a perl lib, which we then use in some places such as calculating the dimensions and swapping them when rending the html markup for a picture. Then we have the thumbor system which has a python library which reads the same exif and in some cases is *different* to the perl lib. So the bug is manifesting itself as the dimensions in html don't match the image dimensions and we end up with a squashed image. BUT in both of those example url's listed above, thumbor is returning an image in the correct orientation, and it is the html markup which is wrong. In other words I don't think this is a thumbor exif bug but a bug in the perl exif parsing lib.

Anyway it seems to me that we could write to very minimal command line scripts that each use both libs and that spit out what they parse as a csv for the entire image collection, then diff em and we've got a list of candidate busted pics to go fix up.

I  don't understand the comment on why we need two versions of the image db?

**Comment by scd** [comment]
I agree that we need more tools to rotate photos.

In this particular instance it is not just rotating, as the aspect ratio is out of sync. The swapped aspect ratio happens when the exif is a particular orientation and thumbor fails. 

I think that thumbor is returning the wrong orientation, but the right dimensions, just the wrong way around. Did you look at the original?

In otherwords a photo which should be displayed as a 5x3 landscape is being returned by thumbor as a 3x5 portrait rotated. Our system thinks thumbor will honor the exif so it sets the dimensions as 5x3. Now we have the situation where a 3x5 image is being resized to fit in a 5x3 frame.

Note that thumbor does output an error message for those photos saying that there is invalid exif format.

Even if we had tools to allow the end user to correct for the problem (ie flip aspect ratio), in a weeks time we may upgrade thumbor and it may suddenly start to work correctly for all exif. In this case all the manual corrections will have to be undone. 

I started to go down the path of having an internal flag for flipping dimensions, but my conclusion was this problem just goes away if we operate thumbor in it's default mode without honoring exif.

I am strongly convinced that we need to change thumbor to operate without honoring exif. If we do that then there is a data migration exercise to add a rotation to photos with exif. To test this in dev I need a version of thumbor which is configured to not honor exif.

This issue screwed with my mind for quite some time. It is definitely not part of this release.

**Comment by brendanheywood** [comment]
I think we are talking about completely different but overlapping and perhaps mutually obfuscating bugs. I'm looking at this page, which only has 2 pics, the left is busted, the right is perfect. Every photo in any photo facet should have a max height of 320 pixels (panos get shrunk but lets ignore that):

https://www.thecrag.com/climbing/australia/arapiles/atridae/area/11962579/photos

The two image links are:

https://static.thecrag.com/img/x320/e3/ba/e3ba54c7e27668c93986f516142ae9cfa97d134d

This one is 180 pixels wide x 320 height

The html is WRONG and served with these dimensions: width="420" height="236.25"

https://static.thecrag.com/img/x320/da/b1/dab183083ba03a78499852e1feb5a4c1816b5f82

This one is 222 wide x 320 high (correct). The html is correct and served with 

```width="221.866666666667" height="320"```

Apart from the weird sized these are both rotated correctly and the aspect is correct for both.

**Comment by brendanheywood** [comment]
Ok so I've found the bug above and fixed in dev, and pushed, it was manipulating the width and height stuff after generating to thumbor url so the changes were being ignored:

https://dev.thecrag.com/climbing/australia/arapiles/atridae/area/11962579/photos

I suspect the topo one is similar, the html is wrong not the image served by thumbor.

**Comment by scd** [comment]
Definitely different bug. Which topo?

The symptoms of the bug which plagued me was a squashed or elongated image

**Comment by brendanheywood** [comment]
I've pasted the link 3 times now! Here is a 4th time: 

https://www.thecrag.com/climbing/australia/arapiles/atridae/area/11962579/photos

vs

https://dev.thecrag.com/climbing/australia/arapiles/atridae/area/11962579/photos

This isn't a topo at all. But I suspect there is a similar bug in the topo sizing code, I'm looking at it now

**Comment by scd** [comment]
I compared those links. I thought you meant another topo not photo. I was going to have a quick look at the topo you were referring to but could not find it :)

Don't worry, no need to discuss this any further.

**Comment by brendanheywood** [comment]
The topo I was referring to is also linked above: https://www.thecrag.com/climbing/canada/sullys-hangout

I'm pretty sure I'll fix that bug too, but now I'm not sure what bug you are referring to, can you provide an example?

**Comment by brendanheywood** [comment]
Ok there is definitely the same bug with topos as what I just fixed, the thumbor image is perfect, the html link to it is busted. I getting lost and pretty frustrated following thumborDynamicImageUrl so I'm calling this a night. I believe it's returning the size of the topo, but this is all backwards, the template needs to already know that size and be passing it into the thumborDynamicImageUrl. All of this logic just seems really wrong to be in the template layer

**Comment by scd** [comment]
See issue #2290 

Actually one of example topos has been removed. I will have to send you the original file because it is easier to test when you follow the upload process.

**Comment by scd** [comment]
I have just had a look at the first topo in sullys hangout. The characteristic of the bug looks like the failing thumbor

https://www.thecrag.com/topo/883126830

https://static.thecrag.com/img/320x/filters:rotate(270)/33/11/3311c7a6e9b6c9c5b0edf3a40eddf4a391b0c416

A 180x320 image is trying to be fit in a 640x360 window. The times 2 scale may be the bug you are referring to but the dimension swapping is the bug I am refering to.

https://static.thecrag.com/img/33/11/3311c7a6e9b6c9c5b0edf3a40eddf4a391b0c416

Thumbor thinks that this image is in landscape mode (ie it is ignoring the EXIF). Open the same document in GIMP and it is portrait mode. 

When you upload the file, our system thinks it is a 2988 x  5312 according to the exif. Our system creates HTML based on this aspect ratio assuming thumbor will display it correctly. We create the HTML viewport dimensions with this aspect ratio. However thumbor incorrectly shows the image with a different rotation.

I am not sure what github will do but I am uploading the image here:

![3311c7a6e9b6c9c5b0edf3a40eddf4a391b0c416 original](https://cloud.githubusercontent.com/assets/373058/21573728/131241d2-cf3c-11e6-8dd1-b89801f2dc6b.jpg)

For your convenience I will email you the file. Be aware that gmail does something funny as well, and may show the image in the wrong rotation. 

Seriously don't waste time code trawling until you understand why the dimensions are flipped and why it is impossible for us to know this. It is not just a matter of the user rotating the image to resync, because rotating rotates both the image and the dimensions. Of course I may still be wrong, but I have invested quite a bit of time into this. This was a very difficult bug to find because it was a regression issue moving from our pre-canned image to thumbor, where there were dozens of other overlapping issues and inconsistent handling of EXIF for various different image software I was using.

**Comment by scd** [comment]
Interestingly git hub is inconsistent with the original within itself. Embedded in the discussion the image has a landscape orientation but click on the image to get the cloud version and it has the portrait orientation.

**Comment by brendanheywood** [comment]
I think a couple more just turned up here:

https://www.thecrag.com/climbing/new-zealand/auckland/area/879854745

eg https://www.thecrag.com/climbing/new-zealand/auckland/area/879854745/topo#t1114924404

**Comment by scd** [comment]
Another:

https://www.thecrag.com/climbing/austria/area/419996787

**Comment by brendanheywood** [comment]
So just circling back to this one and re-testing after the recent change to the thumbor webhelper to always return two dimensions, and not just reply on the browser implicitly resize the photo to the correct size after it has loaded. (as an aside the perceived and real speed impact of this tiny change was well worth it, the browser on a cold cache isn't doing anywhere near as much work re-laying out the page so it feels much snappier. The visual 'speed index' for photos pages is now twice as good in dev even after taking into account the better latency to dev)

Anyway I've long believed that this issue was really like 4 overlapping bugs and I suspect that we've actually chipped away and fixed most of them. Most of the examples in this thread are no longer broken, but I know a couple instances have been manually fixed. If we find more in future we should move them to the test area so they hang around forever.

In any case there is 1 known issues remaining out of the list above, eg on this page there is one photo who aspect is the wrong way around so after the image loads it's vertical dimension is wrong:

https://www.thecrag.com/climbing/australia/arapiles/atridae/area/11962651/photos
![image](https://user-images.githubusercontent.com/187449/28245397-2289860e-6a49-11e7-98c2-90ace6d8ad1d.png)

https://dev.thecrag.com/climbing/australia/arapiles/atridae/area/11962651/photos
![image](https://user-images.githubusercontent.com/187449/28245403-311e4e34-6a49-11e7-9488-22149d974834.png)

Now in dev this same image has a forced both dimensions instead of having a screwy height it instead has  a screwy crop instead. Also you can see that the 2nd and 5th images are now cropped weirdly when in production they were. So now we have 2 bugs instead of 1 :(

BUT but but but ---> when you open up the photo detail page for all 3 of those screwy images then they are *ALL* correct. So this is a bug in our code not thumbor! Originally I thought it was an issue depending on which size we requested, but nope. Then I realized that the two pages json data is quite diverged:

![image](https://user-images.githubusercontent.com/187449/28245634-e01c87ac-6a4e-11e7-954b-7429e7f5645c.png) ![image](https://user-images.githubusercontent.com/187449/28245640-f1d3665a-6a4e-11e7-8f07-9621d0b9bbd1.png)

Also not the origWidth and origHeight are reversed. The two pages call atomPhoto vs dbAtomPhoto. This is where I kinda get a bit lost @scd and am not too keen to touch stuff. It looks like there is a missing call to aspectRotator somewhere (if I add it then it fixes one of the images but not the other 2). 

So I feel like we are finally getting close to as solution for both of these last 2 outstanding issues and close this off

**Comment by brendanheywood** [comment]
btw photo page is this one:

https://dev.thecrag.com/photo/729848244

**Comment by scd** [comment]
There were definitely two bugs which became apparent in the photo facet when we forced both dimensions. 

1. the aspectRotator adjustment function was missing from the facet search results.
2. the sql photo facet was not returning the Exif database settings for each photo.

It was working before because we were only specifying one dimension in the photos. Going to two dimensions meant that the aspect now mattered.

I reviewed the facet code for webcovers and topos and it looks like it includes both the rotator and exif.

Note that the core bug being the thumbor/theCrag interpretation of exif being different for some photos is still not resolved and requires us to either submit a bug report to thumbor (not reading exif info for some photos) or use thumbor without exif.
